### PR TITLE
[Merged by Bors] - Add transform hierarchy stress test

### PR DESCRIPTION
**Pull request by CptPotato**
## Objective

There recently was a discussion on Discord about a possible test case for stress-testing transform hierarchies.

## Solution

Create a test case for stress testing transform propagation.

*Edit:* I have scrapped my previous example and built something more functional and less focused on visuals.

There are three test setups:

- `TestCase::Tree` recursively creates a tree with a specified depth and branch width
- `TestCase::NonUniformTree` is the same as `Tree` but omits nodes in a way that makes the tree "lean" towards one side, like this:
  <details>
  <summary></summary>

  image
  </details>
- `TestCase::Humanoids` creates one or more separate hierarchies based on the structure of common humanoid rigs
  - this can both insert `active` and `inactive` instances of the human rig

It's possible to parameterize which parts of the hierarchy get updated (transform change) and which remain unchanged. This is based on @james7132 suggestion:
There's a probability to decide which entities should remain static. On top of that these changes can be limited to a certain range in the hierarchy (min_depth..max_depth).

### CODE REVIEW

**Comment by superdump** [comment]
Personally I would like to have a home for these kinds of test examples that are useful for development.

**Comment by alice-i-cecile** [comment]
> Personally I would like to have a home for these kinds of test examples that are useful for development.

Seconding this: @CptPotato can you start a top-level "stress-tests" folder for this to live in? We can move the related benchmarky-examples into there in a follow-up PR.

@james7132 @superdump ping me when you think this is ready; I can merge this in for you to help unblock more experiments.

**Comment by james7132** [comment]
A few requests from me:

 - The ability to set the proportion of how many are mutated every frame. This can be used to tell how effective our hierarchy dirtying detection/propagation is. Randomly choosing whether we include or exclude the Rotate component is sufficient. Tentatively, this should probably could also have a control on the minimum and maximum depths this occurs at.
 - Non uniform distributions of hierarchy depths as an optional parameter. We'll probably find a few cases where there are huge numbers of uniform depth transforms, then a few deep cases (thinking active character models among a scene  of otherwise normal scene geometry that might get de-parented/marked static in final builds).

**Comment by CptPotato** [comment]
Good points @james7132. I think I'll also change the code to be more "functional" rather than trying to make it look nice.

I have some ideas for how to address these things. Tough, it might take me a couple of days to get to it.

File: examples/3d/hierarchy_fractal.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+    );
+}
+
+fn main() {
+    App::new()
+        .insert_resource(ClearColor(Color::BLACK))
+        .insert_resource(WindowDescriptor {
+            width: 1400.0,
+            height: 900.0,
+            ..default()
```
Comment (marko-lazic) [inline examples/3d/hierarchy_fractal.rs:209]: I think we should use ..Default::default() here

Comment (CptPotato) [reply examples/3d/hierarchy_fractal.rs:209]: I'm not sure. I think this is fine?

`default()` was recently added and all examples are using it (see #4071)

Comment (marko-lazic) [reply examples/3d/hierarchy_fractal.rs:209]: Oh nice, I forgot to pull :)

**Comment by CptPotato** [comment]
Alright, I created the `stress_test` sub folder and moved the example there.

I have mostly rewritten the example, ditching the focus on visuals and taking some feedback into account, there's some more info in the PR description. The branch name doesn't really fit anymore but I guess that's okay?

I'm still not too sure about the configuration. I tried making it as versatile as reasonably possible but it seems a little bloated to me.

File: examples/stress_tests/transform_hierarchy.rs [hunk 2]
```diff
@@ -0,0 +1,500 @@
...
+        /// the maximum depth
+        depth: u32,
+        /// max number of children per node
+        branch_width: u32,
+    },
+    /// one or multiple humanoid rigs
+    Humanoids {
+        /// number of active instances (uses the specified [UpdateFilter])
+        active: u32,
+        /// numer of inactive instances (always inactive)
```
Comment (james7132) [inline examples/stress_tests/transform_hierarchy.rs:67]:
```suggestion
        /// number of inactive instances (always inactive)
```

**Review by james7132 — APPROVED** [review]
One typo, and please fix the CI, but otherwise amazing. This is exactly the kind of tests we needed.

**Comment by CptPotato** [comment]
Good to hear this is on the right track.

Should we include a number of preset configurations for easier and more consistent testing?
Like:

```bash
cargo r --example transform_hierarchy -- 1
cargo r --example transform_hierarchy -- 2
...
```

**Comment by superdump** [comment]
If we do that, it would be good if they had memorable and descriptive names. 🙂 Like max-depth (a single long chain of hierarchy for worst case), max-width (like the many_cubes — sphere case of 160k root entities with transforms), varied-depth (perhaps specifying a depth range and the branches have paths that are various lengths within that depth range to try to simulate something reasonably realistic)?

**Review by james7132 — CHANGES_REQUESTED** [review]

File: examples/README.md [hunk 3]
```diff
@@ -408,3 +409,9 @@ ruby -run -ehttpd examples/wasm
 To load assets, they need to be available in the folder examples/wasm/assets. Cloning this
 repository will set it up as a symlink on Linux and macOS, but you will need to manually move
 the assets on Windows.
+
+# Stress Tests
+
```
[the remark this answers is not in the document]
Comment (CptPotato) [reply examples/README.md:414]: I added a short description, I'm not really sure what else to write :/

> ...and how to best run them

What exactly do you mean by this? Mentioning that `--release` should be used?

**Comment by CptPotato** [comment]
Sorry for the delay. I added a few test cases that can be run like this:

```sh
cargo r --example transform_hierarchy -- <name>
```

These are the setups I came up with:

| Configuration        | Description                                                       |
| -------------------- | ----------------------------------------------------------------- |
| `large_tree`         | A fairly wide and deep tree.                                      |
| `wide_tree`          | A shallow but very wide tree.                                     |
| `deep_tree`          | A deep but not very wide tree.                                    |
| `chain`              | A chain. 2500 levels deep.                                        |
| `update_leaves`      | Same as `large_tree`, but only leaves are updated.                |
| `update_shallow`     | Same as `large_tree`, but only the first few levels are updated.  |
| `humanoids_active`   | 4000 active humanoid rigs.                                        |
| `humanoids_inactive` | 4000 humanoid rigs. Only 10 are active.                           |
| `humanoids_mixed`    | 2000 inactive and 2000 active humanoid rigs.                      |
| `drawing`            | A rendering-heavy configuration. Meshes and materials are shared. |
| `drawing_unique`     | Same as `drawing`, but all meshes and materials are unique.       |

I shot for ~250k to 350k entities for each of those, I think this should provide a heavy enough load.
The only exception is the `chain` test case. I couldn't make this one much more extreme without running into stack overflows.

**Comment by alice-i-cecile** [comment]
> Edit: I'm thinking about removing the last two configurations (the drawing ones) and removing the related code from the example, unless somebody needs those. Rendering performance seem a little outside of what the stress test is meant for, and it might be a better fit for a separate rendering/batching stress test.

I agree with this.

File: examples/README.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (alice-i-cecile) [reply examples/README.md:414]: Yep, explaining that you should use `--release` should be fine for now IMO.

**Comment by CptPotato** [comment]
Alright, the drawing related code is gone, I also switched to `MinimalPlugins` while I was at it.

I'm fairly happy with this as it is. Let me know if I should add/change anything!

**Review by alice-i-cecile — APPROVED** [review]
Looks good to me! I'm quite happy with this code, and it should be very useful for development.

Once I have another approval I'm happy to merge this in, and then I'd like to reshuffle the other stress tests into this folder.

**Review by james7132 — APPROVED** [review]

**Pull request closed without merging**